File: services/academic-service/src/test/java/com/visor/school/academic/integration/StudentSearchIntegrationTest.java
package com.visor.school.academic.integration;

import com.visor.school.academic.config.TestConfig;
import com.visor.school.academic.model.Student;
import com.visor.school.academic.service.StudentService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@ActiveProfiles("test")
@Import(TestConfig.class)
@Transactional
class StudentSearchIntegrationTest {

    @Autowired
    private StudentService studentService;

    @BeforeEach
    void setup() {
        studentService.enrollStudent(UUID.randomUUID(), "John", "Doe", LocalDate.of(2010, 1, 1), 5, null, null);
        studentService.enrollStudent(UUID.randomUUID(), "Jane", "Smith", LocalDate.of(2010, 6, 1), 5, null, null);
        studentService.enrollStudent(UUID.randomUUID(), "Bob", "Johnson", LocalDate.of(2006, 1, 1), 12, null, null);
    }

    @Test
    void shouldSearchStudentsByName() {
        List<Student> students = studentService.searchStudentsByName("John");
        assertFalse(students.isEmpty());
    }

    @Test
    void shouldGetStudentsByGradeLevel() {
        List<Student> grade5Students = studentService.getStudentsByGradeLevel(5);
        assertTrue(grade5Students.size() >= 2);
        assertTrue(grade5Students.stream().allMatch(s -> s.getGradeLevel() == 5));
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/integration/StudentSearchIntegrationTest.kt
package com.visor.school.academicservice.integration

import com.visor.school.academicservice.model.EnrollmentStatus
import com.visor.school.academicservice.repository.StudentRepository
import com.visor.school.academicservice.service.StudentService
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.util.UUID

/**
 * Integration test for student search functionality
 */
@SpringBootTest
@ActiveProfiles("test")
@Transactional
class StudentSearchIntegrationTest @Autowired constructor(
    private val studentService: StudentService,
    private val studentRepository: StudentRepository
) {

    @BeforeEach
    fun setup() {
        // Create test students
        val userId1 = UUID.randomUUID()
        val userId2 = UUID.randomUUID()
        val userId3 = UUID.randomUUID()

        studentService.enrollStudent(
            userId = userId1,
            firstName = "John",
            lastName = "Doe",
            dateOfBirth = LocalDate.of(2010, 1, 1),
            gradeLevel = 5
        )

        studentService.enrollStudent(
            userId = userId2,
            firstName = "Jane",
            lastName = "Smith",
            dateOfBirth = LocalDate.of(2010, 6, 1),
            gradeLevel = 5
        )

        studentService.enrollStudent(
            userId = userId3,
            firstName = "Bob",
            lastName = "Johnson",
            dateOfBirth = LocalDate.of(2006, 1, 1),
            gradeLevel = 12
        )
    }

    @Test
    fun `should search students by first name`() {
        // When
        val results = studentService.searchStudentsByName("John")

        // Then
        assertTrue(results.isNotEmpty())
        assertTrue(results.any { it.firstName.contains("John", ignoreCase = true) })
    }

    @Test
    fun `should search students by last name`() {
        // When
        val results = studentService.searchStudentsByName("Smith")

        // Then
        assertTrue(results.isNotEmpty())
        assertTrue(results.any { it.lastName.contains("Smith", ignoreCase = true) })
    }

    @Test
    fun `should search students by full name`() {
        // When
        val results = studentService.searchStudentsByName("John Doe")

        // Then
        assertTrue(results.isNotEmpty())
        assertTrue(results.any { 
            it.firstName.contains("John", ignoreCase = true) && 
            it.lastName.contains("Doe", ignoreCase = true) 
        })
    }

    @Test
    fun `should return empty list for non-existent student`() {
        // When
        val results = studentService.searchStudentsByName("Nonexistent")

        // Then
        assertTrue(results.isEmpty())
    }

    @Test
    fun `should find students by grade level`() {
        // When
        val grade5Students = studentService.getStudentsByGradeLevel(5)
        val grade12Students = studentService.getStudentsByGradeLevel(12)

        // Then
        assertTrue(grade5Students.size >= 2)
        assertTrue(grade5Students.all { it.gradeLevel == 5 })
        assertTrue(grade12Students.size >= 1)
        assertTrue(grade12Students.all { it.gradeLevel == 12 })
    }

    @Test
    fun `should find students by enrollment status via repository`() {
        // When - Access repository directly for enrollment status search
        val enrolledStudents = studentRepository.findByEnrollmentStatus(EnrollmentStatus.ENROLLED)

        // Then
        assertTrue(enrolledStudents.isNotEmpty())
        assertTrue(enrolledStudents.all { it.enrollmentStatus == EnrollmentStatus.ENROLLED })
    }

    @Test
    fun `should find student by student ID`() {
        // Given
        val userId = UUID.randomUUID()
        val student = studentService.enrollStudent(
            userId = userId,
            firstName = "Find",
            lastName = "ByID",
            dateOfBirth = LocalDate.of(2010, 1, 1),
            gradeLevel = 5
        )

        // When
        val found = studentService.getStudentByStudentId(student.studentId)

        // Then
        assertNotNull(found)
        assertEquals(student.studentId, found?.studentId)
        assertEquals("Find", found?.firstName)
    }
}

